File: leetcode/67.AddBinary.go
package leetcode

import (
	"math"
	"strconv"
)

func pad(s string, pad string, length int) string {
	if len(s) >= length {
		return s
	}

	for i := len(s); i < length; i++ {
		s = pad + s
	}
	return s
}

func addBinary(a string, b string) string {
	result := ""
	carry := 0
	max := int(math.Max(float64(len(a)), float64(len(b))))
	padA := pad(a, "0", max)
	padB := pad(b, "0", max)

	for i := max - 1; i >= 0; i-- {
		numA, _ := strconv.Atoi(string(padA[i]))
		numB, _ := strconv.Atoi(string(padB[i]))

		total := numA + numB + carry
		char := strconv.Itoa(total % 2)
		result = char + result
		carry = int(math.Floor(float64(total / 2)))
	}

	if carry != 0 {
		result = "1" + result
	}

	return result
}

File: leetcode/67.AddBinary.ts
function addBinary(a: string, b: string): string {
  let result = ''
  let carry = 0
  const max = Math.max(a.length, b.length)
  const padA = a.padStart(max, '0')
  const padB = b.padStart(max, '0')

  for (let i = max - 1; i >= 0; i--) {
    const numA = Number(padA[i] ?? 0)
    const numB = Number(padB[i] ?? 0)

    const total = numA + numB + carry
    const char = String(total % 2)
    result = char + result
    carry = Math.floor(total / 2)
  }

  if (carry) result = '1' + result

  return result
}
console.log(addBinary('11', '1'))
console.log(addBinary('1010', '1011'))
